Build bounding boxes properly

andrsd opened the pull request:
BoundingBox class in libMesh assumes that x-, y- and z- coordinates
are monotonically increasing. "Sort" the coordinates before constructing
the bounding, so that users do not have to think about this.

Refs #1275

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

loganharbour commented:
Searching for `"bottom_left"` looks like there's a few other methods that could probably benefit from something similar...

andrsd commented:
So, `MooseBoundingBox`?

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/515364/) on 05bb664 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15187/site)

This comment will be updated on new commits.

loganharbour commented:
> So, `MooseBoundingBox`? 😄

I see nothing wrong with this :grimacing:

andrsd commented:
I will try to get this in libMesh and if they refuse it, we can do something in MOOSE.

loganharbour commented:
> I will try to get this in libMesh and if they refuse it, we can do something in MOOSE.

Great. I would say some method on the `BoundingBox` that is called after it is created to make it positive if it isn't.

andrsd commented:
I am actually changing the c-tors :-)

andrsd commented:
Closing in favor of #15192

andrsd commented:
I will re-purpose this and fix the other instances where we build bounding boxes...

andrsd committed:
Properly construct bounding boxes

Users should not need to know that libMesh assumes sorted coordinates of
bounding boxes. Any valid combination of 2 corners should build a correct
bounding box.

Closes #15208

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/517651/) on ace37fc wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/15187/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/15187/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1afbd334ab37341014d20a00abf801ce474276c2`

loganharbour commented on framework/include/utils/MooseUtils.h:
```diff
@@ -778,6 +779,15 @@ struct IsLikeReal<DualReal>
 };
 
+/**
+ * Construct a valid bounding box from 2 arbitrary points
```
Can you you put a comment in here as to _why_ this is the preferred method to build a bounding box instead of using the standard constructor?

If you can think of another method name that would suggest that this picks the correct orientation, that would be great too. But, I'm struggling to come up with one so it's not necessary.

andrsd replied:
> Can you you put a comment in here as to why this is the preferred method to build a bounding box instead of using the standard constructor?

Yeah, I was thinking about that. But, not all bounding boxes in the framework are/should be constructed this way. So, I will add _when_ and also _why_ to use this 👍 

> If you can think of another method name that would suggest that this picks the correct orientation, that would be great too. But, I'm struggling to come up with one so it's not necessary.

I don't think the name should suggest anything about the orientation at all. It is an implementation detail of the `BoundingBox` class. You as a user want a bounding box from 2 points. Why do you need to care about how it is done inside the class, right? The explaining comment should be enough here...
